### PR TITLE
feat: source configuration files

### LINKED ISSUE #155: good auto config options
look for
* [x] requirements.txt (https://github.com/gakonst/foundry/pull/181)
* [ ] .dapprc
* [ ] .forgerc

under:
* the current directory
* `$HOME/`
* `$HOME/.foundry`
* `$HOME/.config/foundry`

### PR DESCRIPTION
**Pull request by odyslam**
- fixes #155 
- Doesn't work, because `source` doesn't propagate but works for the subshell. On top, while `UNIX` commands like `cat`, run properly and find the config files, `source` outputs an error.

### CODE REVIEW

File: cli/src/forge.rs [hunk 1]
```diff
@@ -27,7 +28,39 @@ mod utils;
 #[tracing::instrument(err)]
 fn main() -> eyre::Result<()> {
     utils::subscriber();
-
+    if cfg!(target_os = "windows") {
+        println!("Can't source env files on windows");
+    } else {
+        let home: String = env::var("HOME")?;
```
Comment (mattsse) [inline cli/src/forge.rs:34]: there is the https://crates.io/crates/dirs crate that has `dirs::home_dir()`

File: cli/src/forge.rs [hunk 2]
```diff
@@ -27,7 +28,39 @@ mod utils;
...
+            PathBuf::from(&home),
+            PathBuf::from(&home).join(".foundry"),
+            PathBuf::from(&home).join(".config/foundry"),
+        ];
+        let files = [String::from(".dapprc"), String::from(".forgerc")];
+        let mut flag = false;
+        for path in paths.iter() {
+            for file in &files {
+                let arg = PathBuf::from(&path).join(file);
+                match std::process::Command::new("source").args(arg.to_str()).spawn() {
```
Comment (mattsse) [inline cli/src/forge.rs:45]: isn't source a bash command? so we would need to run `bash -c source {}` instead?

Comment (odyslam) [reply cli/src/forge.rs:45]: Yeah, sorry about not putting the whole context. So, apparently, the correct command is `. .dapprc` in order to export the env variables when called from an external program. 

The issue that I have (for unknown reasons still) is that while `cat` will work as expected, finding the source file and showing it's content, `source` and `.` will output a `file not found` error. 

I will have to dig in this. A solution we thought we @gakonst is perhaps using a crate where we can define a bash script as a literal string and execute that script, which will do the exports for us. 

Let me do the investigation tonight or tomorrow morning and come back with results.

File: cli/src/forge.rs (hunk 1 above)
Comment (odyslam) [reply cli/src/forge.rs:34]: yy, I generally try to follow a minimal crate approach. Do you think that using the HOME env variable has gotchas (for UNIX systems that is). 

For windows, I am not sure we even support the platform, that's why I made the distinction.

**Comment by odyslam** [comment]
The env variables are not accessible by the main rust program, but is set correctly in the bash script. 

We could output the env variables via `env | grep DAPP*` and have them in the `output` string. Then we can safely parse them. ?

**Comment by odyslam** [comment]
@gakonst I think I found a good solution. The command will output the env variables in the file, after sourcing it as expected. It's trivial to read it then by line and split it. 

The env variable is then set by the rust program, thus making it available for later use. 

I wasn't able to test it with a project that I currently work on due to some remappings issues (dapptools OG work OK), but I tested it with printing the env variable after the soucing part,  in the rust program.

**Review by mattsse — CHANGES_REQUESTED** [review]
If I understand that correctly, this basically sets some env vars manually that are then accessible in the general program.

I'm afraid running an additional bash script will be rather hard to maintain and troubleshoot.

I think what we want is a simple way to get access to all import config params, like a `Config` struct.

Eventually, we'd like to support `~/.foundry.toml` so what we should aim for is something like

```rust
fn get_config() -> Config {
  let mut config = read_config_toml("~/.foundry").unwrap();
  
  // replace all config values with env vars if present
  // --snip--
  config
}
```

so first read the config toml, env can override those values

**Comment by odyslam** [comment]
The use-case for this PR is to make it work as a drop-in replacement with dapptools. I totally agree that we should roll out our own `.toml` configuration as it's much saner.

This insanity that I coded is the only way I could find in order to `source` a bash script and make the env variables available inside the rust program (which is the parent of the subshell that sources the config).

**Comment by gakonst** [comment]
I don't think we'll be going with this approach, given how hard it's been to source a file. For now, people that want to use `.dapprc` can always `source .dapprc` or `. .dapprc`.

We continue the discussion in #291 around a foundry.toml. 

If anyone feels strongly about it and can drive this feature over the line, feel free to go for it.

**Pull request closed without merging**